doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,274 @@
+3. `Q_final` will contain the temporal predicate `lower_ts <= mz_now()` and a projection that hides
+   the last two columns, which contain the two timestamps of the validity period. This will ensure
+   that records are only inserted at the lower bound of their validity period and that `Q_expr` will
+   see rows with the correct schema.
+
+Given this decomposition we can now define the ingestion as:
+
+1. `C_intermediate = PERSIST(Q_expiration(Q_safemfp(RAW)))`
+2. `C = PERSIST(Q_expr(Q_final(C_intermediate)))`
+
+With these equations we have accomplished the goal of doing as much work as possible before the
+first persist operator while requiring no past state from `RAW` to resume. To see why that is the
+case, let's describe the state needed by the `Q_expiration` operator. The state of `Q_expiration` at
+time `t` is the set of records `(row, t_record, diff)` for which `t_record <= t && t <
+upper_ts(Q_mfp(row))`. In other words, for every timestamp it needs to remember all past records
+that expire after that timestamp.
+
+If we now look at the output of `Q_safemfp` we see that it converts a `(row, t_record, diff)` update
+to `((row, lower_ts, upper_ts), t_record, diff)` update. Additionally, `Q_expiration` will transform
+each `((row, lower_ts, upper_ts), t_record, diff)` update into two:
+
+1. `((row, lower_ts, upper_ts), t_record, diff)`
+2. `((row, lower_ts, upper_ts), upper_ts, -diff)`
+
+Given the above we can see that at the beginning of an ingestion execution `Q_expiration` can
+recover its in-memory state at time `t` by taking a snapshot of `C_intermediate` (not `RAW`) at time
+`t`, which is available because it is stored in persist! This is because any record that may expire
+in the future is maintained until the end of its validity period and `Q_mfp` has preserved the
+previously computed timestamps. This technique is similar to how the `UPSERT` envelope recovers its
+state from its output.
+
+### Planner/Optimizer/Data structure stabilization
+
+The inability to have `RAW` snapshots readily available has some deep implications on how much of
+our internal choices we are willing to stabilize, which in turn have deep implications on the kind
+of queries we want to allow users to use in the `TRANSFORM USING` position.
+
+In the equations above we had to define `C_intermediate` to effectively be a collection whose schema
+is very much dependent on planner and optimizer choices. Since the transformation from `RAW` to
+`C_intermediate` is not described by high level `SQL` that can be stored in the catalog we must
+explore what else we must stabilize in order to be able to continue ingesting a source across
+versions.
+
+There are various flavors of how this could go. I will lay out some of them below.
+
+**Option 1: Only accept transforms `Q` whose `Q_expr` is the identity, stored as SQL**
+
+We make the guarantee that if a version of Materialize plans and optimizes a piece of SQL to just an
```

Comment:
This is similar to what we do in the catalog for [materialized] view definitions today so I think it should be fine here as well. The additional stability constraint about MFP also seems reasonable under the caveat the we might produce different MFPs across restarts (for example if the optimizer to learn how to better simplify complex predicates).